### PR TITLE
Calico aws

### PR DESCRIPTION
**Pull request by naumvd95**
WIP Define parameters  overriding for Jenkins

    Use cases:
    1) Calico managing in case of using mcp-job calico-ha-aws-k8s
    2) overriding parameters in case of any jenkins job

    From UI part:
    At least need to provide possibility of defining key(str):value(str) parameters
scheme = https://www.lucidchart.com/invitations/accept/6e2bbc84-a5df-43f9-8b21-2a5b754a57d8

### CODE REVIEW

File: kqueen/engines/jenkins.py [hunk 1]
```diff
@@ -64,6 +81,21 @@ def __init__(self, cluster, **kwargs):
         self.client = self._get_client()
         # Cache settings
         self.cache_timeout = 5 * 60
+        self.job_override_params = kwargs.get('override_parameters', {})
+        logger.critical('OVERRIDEDEEE {}'.format(self.job_override_params))
```
Comment (katyafervent) [inline kqueen/engines/jenkins.py]: :sweat_smile:

File: kqueen/engines/jenkins.py [hunk 2]
```diff
@@ -74,39 +106,40 @@ def _get_provision_job_builds(self):
...
-        try:
-            client = jenkins.Jenkins(config.get('JENKINS_API_URL'), **conn_kw)
-            auth_verify = client.get_whoami()
-            if auth_verify:
-                status = config.get('PROVISIONER_OK_STATE')
-        except Exception as e:
-            logger.exception('Could not contact JenkinsEngine backend: ')
-            status = config.get('PROVISIONER_ERROR_STATE')
-        return status
+        Check, that defined key parameter exist in Job
```
Comment (katyafervent) [inline kqueen/engines/jenkins.py]: that defined key parameter exist in Job -> that key parameter, defined by the user exists in the specified Job

File: kqueen/engines/jenkins.py [hunk 3]
```diff
@@ -74,39 +106,40 @@ def _get_provision_job_builds(self):
...
-            :obj:`jenkins.Jenkins`: initialized Jenkins client
+            bool: True if all parameters exist
         """
-        return jenkins.Jenkins(self.jenkins_url, **{
-            'username': self.username,
-            'password': self.password,
-            'timeout': 10
-        })
+
+        job_body = self.client.get_job_info(self.provision_job_name, depth=1)
```
Comment (katyafervent) [inline kqueen/engines/jenkins.py]: I suggest to add try-catch here and return False from function end at catch block

File: kqueen/engines/jenkins.py [hunk 4]
```diff
@@ -118,6 +151,17 @@ def provision(self, **kwargs):
         # PATCH THE CTX TO CONTAIN CLUSTER NAME AND UUID
         ctx[cluster_name] = 'kqueen-{}'.format(self.cluster.id)
         ctx[cluster_uuid] = self.cluster.id
+        if self.job_override_params:
```
Comment (katyafervent) [inline kqueen/engines/jenkins.py]: I suggest to manage all job parameters in a one place, so first let's add cluster_name and cluster_uuid to self.job_override_params (also for me self.job_parameters sounds better) and then manage everything in the same way

File: kqueen/engines/jenkins.py [hunk 5]
```diff
@@ -118,6 +151,17 @@ def provision(self, **kwargs):
...
+            for key, value in self.job_override_params:
```
Comment (katyafervent) [inline kqueen/engines/jenkins.py]: self.job_override_params.items()

File: kqueen/engines/jenkins.py [hunk 6]
```diff
@@ -118,6 +151,17 @@ def provision(self, **kwargs):
...
+                ctx[key] = ctx[value]
+
+        for param in ctx.keys():
+            if not self._parameter_exist(param):
+                msg = 'Failed to manage job parameter: {}'.format(param)
+                logger.error(msg)
+                return False, msg
+
+        logger.debug('Current Job configuration: {}, provisioning started...'.format(ctx))
```
Comment (katyafervent) [inline kqueen/engines/jenkins.py]: actually, it is started after 166 line

File: kqueen/engines/jenkins.py (hunk 4 above)
Comment (naumvd95) [reply kqueen/engines/jenkins.py]: imo its different cases - override-parameters used for overriding sane-default parameters, like boolean variables, checkers, additional cases
cluster name and UUID - must have cases, and it doesnot have any sane-defaults, so i prefer to manage it from separate UI-fields

File: kqueen/engines/jenkins.py [hunk 7]
```diff
@@ -47,9 +47,26 @@ class JenkinsEngine(BaseEngine):
                 'validators': {
                     'required': True
                 }
+            },
+            # TODO below form should be increased dynamically in case of 'add one more parameter'
+            # Use-case: 'add parameter'->'key': 'value'-> 'add parameter'-> etc...
+            # Should return dict{}
+            'override_parameters': {
```
Comment (katyafervent) [inline kqueen/engines/jenkins.py]: please, replace with
            'override_parameters': {
                'type': 'parameters',
                'label': 'Jenkins job parameters',
                'order': 2,
                'validators': {
                          'required': False
                }
            }
        },
        'cluster': {}
    }

Also, did you delete clusters occasionally?

File: kqueen/engines/jenkins.py (hunk 6 above)
Comment (naumvd95) [reply kqueen/engines/jenkins.py]: there are try catch block, so i prefer to log here, its bad?

Comment (katyafervent) [reply kqueen/engines/jenkins.py]: it's better to place it inside try-catch block

**Comment by naumvd95** [comment]
closed as duplicate

**Pull request closed without merging**